### PR TITLE
feat: improve imagenet gaea example [DET-3319]

### PR DESCRIPTION
**Pull request by liamcli**
## Description
Add tricks to improve performance for the ImageNet GAEA examples:
- Switch to RandAugment from AutoAugment
- Add drop_path regularization
- Add EMA moving average for validation
- Change training routine to be more similar to that used for EfficientNet (https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet).

## Test Plan
- Running experiment now to test convergence result.

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 1]
```diff
@@ -11,25 +11,121 @@
...
+    accuracy,
+    AvgrageMeter,
+)
 
 Genotype = namedtuple("Genotype", "normal normal_concat reduce reduce_concat")
 
 activation_map = {"relu": nn.ReLU, "swish": Swish, "hswish": HSwish}
 
 
+class LinearLRScheduler(_LRScheduler):
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:41]: Question: Does this need a save_dict function? i.e. was this tested that the pause and reload functionally saves these values correctly?

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 2]
```diff
@@ -11,25 +11,121 @@
...
+        self.optimizer = optimizer
+        self.warmup_epochs = warmup_epochs
+        self.max_epochs = max_epochs
+        self.last_epoch = last_epoch
+        super(LinearLRScheduler, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            lr_mult = (self.last_epoch + 1) / self.warmup_epochs
+        elif self.max_epochs - self.last_epoch > self.warmup_epochs:
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:52]: Blocking: can you add some comments for what it's checking for?

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 3]
```diff
@@ -11,25 +11,121 @@
...
+    def _get_closed_form_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            return [
+                (self.last_epoch + 1) / self.warmup_epochs * b_lr
+                for b_lr in self.base_lrs
+            ]
+        return super(CosineWrapper, self)._get_closed_form_lr()
+
+
+class LinearWrapper(LinearLRScheduler):
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:105]: lets move all these LRSchedulers to a new file.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 4]
```diff
@@ -208,3 +316,26 @@ def evaluate_batch(self, batch: Any, model: nn.Module) -> Dict[str, Any]:
         top1, top5 = accuracy(logits, target, topk=(1, 5))
 
         return {"loss": loss, "top1_accuracy": top1, "top5_accuracy": top5}
+
+    # This function is never called by PyTorch Trial but can be used manually
+    # to evaluate model on full validation dataset.
+    def evaluate_epoch(
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:322]: Question: little confused about why we are adding this? 1. how does this differ from evaluate_batch? If we want to use a full validation dataset, we should just use evaluate_full_dataset(). However, part of the perks for Determined is the user doesn't need to keep track of having to average work.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 5]
```diff
@@ -11,25 +11,121 @@
...
+
+    def get_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            return [
+                (self.last_epoch + 1) / self.warmup_epochs * b_lr
+                for b_lr in self.base_lrs
+            ]
+        return super(CosineWrapper, self).get_lr()
+
+    def _get_closed_form_lr(self):
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:96]: blocking: This looks doubled as get_lr. I believe we can just have get_lr since it's being treated as a get_closed_form_lr anyways.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 6]
```diff
@@ -11,25 +11,121 @@
...
+                (self.last_epoch + 1) / self.warmup_epochs * b_lr
+                for b_lr in self.base_lrs
+            ]
+        return super(LinearWrapper, self).get_lr()
+
+    def _get_closed_form_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            return [
+                (self.last_epoch + 1) / self.warmup_epochs * b_lr
+                for b_lr in self.base_lrs
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:124]: same as above

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py (hunk 5 above)
Comment (katport) [reply examples/experimental/imagenet_nas_arch_pytorch/model_def.py]: Lets then set a helper function that handles this inner if statement to prevent duplicate code.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py (hunk 4 above)
Comment (katport) [reply examples/experimental/imagenet_nas_arch_pytorch/model_def.py]: evaluate_batch still calls the full validation dataset. Is there a drastic time difference between using the evaluate_full_dataset and the evaluate_batch?

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py (hunk 5 above)
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/model_def.py]: i created a lr warmup wrapper for any lr scheduler.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py (hunk 4 above)
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/model_def.py]: i learned that both evaluate_batch and evaluate_full_dataset loop through the entire validation with the difference being evaluate_full_dataset gives more control at the cost of losing parallelism

File: examples/experimental/trial/imagenet_nas_arch_pytorch/const.yaml [hunk 7]
```diff
@@ -59,7 +70,7 @@ hyperparameters:
 resources:
   # We recommend 8 slots with V100
   # and at least 16 slots with K80 GPUs.
-  slots_per_trial: 1
+  slots_per_trial: 2
```
Comment (katport) [inline examples/experimental/trial/imagenet_nas_arch_pytorch/const.yaml:73]: Question: why two here? It's a const.yaml so i would expect slots_per_trial: 1. If it's due to memory, you can set it to one and put a note in the readme and here the required space.

File: examples/experimental/trial/imagenet_nas_arch_pytorch/distributed_darts.yaml [hunk 8]
```diff
@@ -31,40 +37,52 @@ data:
...
+checkpoint_storage:
+  save_experiment_best: 5
+  save_trial_best: 2
+  save_trial_latest: 1
+
 searcher:
   name: single
   metric: top1_accuracy 
-  max_steps: 300 # 1.2 million images in training set of imagenet
+  max_steps: 1000 # 1.2 million images in training set of imagenet
```
Comment (katport) [inline examples/experimental/trial/imagenet_nas_arch_pytorch/distributed_darts.yaml:85]: why the difference in max_steps between const and dtrain?

File: examples/experimental/trial/imagenet_nas_arch_pytorch/const.yaml (hunk 7 above)
Comment (liamcli) [reply examples/experimental/trial/imagenet_nas_arch_pytorch/const.yaml]: changed back to 1

File: examples/experimental/trial/imagenet_nas_arch_pytorch/distributed_darts.yaml (hunk 8 above)
Comment (liamcli) [reply examples/experimental/trial/imagenet_nas_arch_pytorch/distributed_darts.yaml:85]: this is to train to convergence.  const uses fake data.
